interfaces/cython/cantera/utils.pyx

```diff
@@ -117,5 +125,209 @@ cdef anyvalue_to_python(string name, CxxAnyValue& v):
 
 cdef anymap_to_dict(CxxAnyMap& m):
     m.applyUnits()
+    if m.isEmpty():
+        return {}
     return {pystr(item.first): anyvalue_to_python(item.first, item.second)
             for item in m.ordered()}
+
+cdef CxxAnyMap dict_to_anymap(data) except *:
+    cdef CxxAnyMap m
+    for k, v in data.items():
+        m[stringify(k)] = python_to_anyvalue(v, k)
+    return m
+
+cdef get_types(item):
+    """ Helper function used by python_to_anyvalue """
+    if not len(item):
+        # Empty list, so no specific type can be inferred
+        return None, 1
+
+    elif isinstance(item, np.ndarray):
+        if isinstance(item.flat[0], numbers.Integral):
+            itype = numbers.Integral
+        elif isinstance(item.flat[0], numbers.Real):
+            itype = numbers.Real
+        else:
+            itype = item.dtype.type
+        return itype, item.ndim
+
+    else:
+        itype = set()
+        for i in item:
+            if isinstance(i, dict):
+                itype.add(dict)
+            elif isinstance(i, list) or isinstance(i , tuple):
+                itype.add(list)
+            elif type(i) == bool:
+                itype.add(bool)  # otherwise bools will get counted as integers
+            elif isinstance(i, numbers.Integral):
+                itype.add(numbers.Integral)
+            elif isinstance(i, numbers.Real):
+                itype.add(numbers.Real)
+            else:
+                itype.add(type(i))
+
+        if itype == {numbers.Real, numbers.Integral}:
+            itype = {numbers.Real}
```

Comment:
2 questions here:

1. Are the first two conditions needed? It seems as though `itype.add(type(i))` would do well enough for those.
2. For the numbers, could you combine that condition? Since you reduce the set to `numbers.Real` at the end anyways. Or is there a use to having just `numbers.Integral` as the type?

It might be worth adding comments that answer these questions, for any future readers.

Reply:
I can see from code below that the answer to question 2 is, yes there is a use to having `numbers.Integral` as the type.

Reply:
Ad (1) ... I don't believe that the two conditions can be merged. Of course the second one can be simplified to `isinstance(i, (list, tuple))`, but that's code golf ...

Reply:
@ischoegl is right that `list` + `tuple` check can be combined.

The reason for not just using `itype.add(type(i))` is that I want to treat any type that inherits from `dict` as a `dict`, and likewise any types inheriting from `list` or `tuple` are essentially `list`-like for the purpose of serialization. I can add comments to this effect if you think it would help.